### PR TITLE
add catboost integration

### PR DESCRIPTION
**Pull request by tohmae**
This PR addresses issue [optuna/optuna#2464](https://github.com/optuna/optuna/issues/2464) and PR [optuna/optuna#2734](https://github.com/optuna/optuna/pull/2734).

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@nzw0301 Could you review this PR since you have already involved in 2734. If you don't have time, please feel free to remove the assignness.

**Comment by Crissman** [comment]
@tohmae Can you add tests for this in `.github/workflows`?
Also, please add to the README.md.

@nzw0301 This has been moved out of draft. Can you take a look at this?

**Comment by nzw0301** [comment]
@Crissman Sure. But I'm waiting for a feedback from catboost developer team to remove this part
https://github.com/optuna/optuna-examples/pull/27/files#diff-773768b8684047a9971f94d174cc68d1dbc2cef09fc95855598b5c641af31af0R55;  more details: https://github.com/optuna/optuna/pull/2734#discussion_r651524382

File: catboost/catboost_integration.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        eval_set=[(valid_x, valid_y)],
+        verbose=0,
+        early_stopping_rounds=100,
+        callbacks=[pruning_callback],
+    )
+    pruning_callback.check_pruned()
+
+    preds = gbm.predict(valid_x)
+    pred_labels = np.rint(preds)
+    accuracy = accuracy_score(valid_y, pred_labels)
```
Comment (nzw0301) [inline catboost/catboost_integration.py:59]: How about using caboost's evaluation function `eval-metrics`?
https://catboost.ai/docs/concepts/python-reference_catboost_eval-metrics.html

File: catboost/catboost_integration.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    return accuracy
```
Comment (nzw0301) [inline catboost/catboost_integration.py:61]: Maybe silly comment: In my understanding, AUC score on the validation dataset should be returned for consistency to the training objective. Otherwise, how about optimizing the accuracy of the training dataset. What do you think?

File: catboost/catboost_integration.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import catboost as cb
+from sklearn.datasets import load_breast_cancer
+from sklearn.metrics import accuracy_score
+from sklearn.model_selection import train_test_split
+
+
+def objective(trial):
+    data, target = load_breast_cancer(return_X_y=True)
+    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.3)
```
Comment (nzw0301) [inline catboost/catboost_integration.py:24]: [minor comment] How about using the same `test_size` as [`lightgbm_integration.py`](https://github.com/optuna/optuna-examples/blob/6a6b20ad634627eebb3e7e104f73b70b45c6e624/lightgbm/lightgbm_integration.py) for consistency? Maybe we can address this part as another PR. (cc: @Crissman) 

```suggestion
    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
```

File: catboost/catboost_integration.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+from sklearn.metrics import roc_auc_score
+from sklearn.model_selection import train_test_split
+
+
+class CatBoostPruningCallback(object):
+    """Callback for catboost to prune unpromising trials.
+
+    See `the example <https://github.com/optuna/optuna-examples/blob/main/
+    catboost/catboost_integration.py>`__
+    if you want to add a pruning callback which observes validation accuracy.
```
Comment (nzw0301) [inline catboost/catboost_integration.py:28]:
```suggestion
```

File: catboost/catboost_integration.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    You must call check_pruned after training unlike other pruning callbacks.
```
Comment (nzw0301) [inline catboost/catboost_integration.py:30]:
```suggestion
    You must call `check_pruned` after training manually unlike other pruning callbacks.
```

File: catboost/catboost_integration.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        self, trial: optuna.trial.Trial, metric: str, valid_name: str = "validation"
+    ) -> None:
+        self._trial = trial
+        self._metric = metric
+        self._valid_name = valid_name
+        self._pruned = False
+        self._message = ""
+
+    def after_iteration(self, info: Any) -> bool:
+        """Run after each iteration."""
```
Comment (nzw0301) [inline catboost/catboost_integration.py:60]: Could add an explanation of `info` object?

File: catboost/catboost_integration.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        if self._metric not in metrics:
+            raise ValueError(
+                'The entry associated with the metric name "{}" '
+                "is not found in the evaluation result list {}.".format(self._metric, info)
+            )
+        scores = metrics[self._metric]
+        current_score = scores[-1]
+        self._trial.report(current_score, step=step)
+        if self._trial.should_prune():
+            self._message = "Trial was pruned at iteration {}.".format(step)
```
Comment (nzw0301) [inline catboost/catboost_integration.py:77]: Is this message used in this example?

File: catboost/catboost_integration.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            self._pruned = True
+            return False
+        return True
+
+    def check_pruned(self) -> None:
+        """Check whether pruend."""
+        if self._pruned is True:
```
Comment (nzw0301) [inline catboost/catboost_integration.py:84]: Equivalent to the following line
```suggestion
        if self._pruned
```

File: catboost/catboost_integration.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            raise optuna.TrialPruned(self._message)
+
+
+def objective(trial):
+    data, target = load_breast_cancer(return_X_y=True)
+    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
+
+    param = {
+        "objective": trial.suggest_categorical("objective", ["Logloss", "CrossEntropy"]),
+        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.01, 0.1),
```
Comment (nzw0301) [inline catboost/catboost_integration.py:94]:
```suggestion
        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.01, 0.1, log=True),
```

File: catboost/catboost_integration.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            "bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]
+        ),
+        "used_ram_limit": "3gb",
+        "eval_metric": "AUC",
+    }
+
+    if param["bootstrap_type"] == "Bayesian":
+        param["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
+    elif param["bootstrap_type"] == "Bernoulli":
+        param["subsample"] = trial.suggest_float("subsample", 0.1, 1)
```
Comment (nzw0301) [inline catboost/catboost_integration.py:107]:
```suggestion
        param["subsample"] = trial.suggest_float("subsample", 0.1, 1, log=True)
```

File: catboost/catboost_integration.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    gbm.fit(
+        train_x,
+        train_y,
+        eval_set=[(valid_x, valid_y)],
+        verbose=0,
+        early_stopping_rounds=100,
+        callbacks=[pruning_callback],
+    )
+
+    pruning_callback.check_pruned()
```
Comment (nzw0301) [inline catboost/catboost_integration.py:121]:
```suggestion
    # evoke pruning manually.
    pruning_callback.check_pruned()
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Let me leave minor comments

File: catboost/catboost_integration.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            return False
+        return True
+
+    def check_pruned(self) -> None:
+        """Check whether pruend."""
+        if self._pruned is True:
+            raise optuna.TrialPruned(self._message)
+
+
+def objective(trial):
```
Comment (nzw0301) [inline catboost/catboost_integration.py:88]:
```suggestion
def objective(trial: optuna.Trial) -> float:
```

**Comment by tohmae** [comment]
Thank you for your comments.
I reflected your suggestions.

**Comment by nzw0301** [comment]
Sorry for the delayed response. I'll leave a comment this week.

best regards,

File: catboost/catboost_pruning.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
+"""
+Optuna example that demonstrates a pruner for CatBoost.
+
+In this example, we optimize the validation accuracy of cancer detection using CatBoost.
+We optimize both the choice of booster model and their hyperparameters. Throughout
```
Comment (nzw0301) [inline catboost/catboost_pruning.py:5]:
```suggestion
We optimize both the choice of booster models and their hyperparameters. Throughout
```

File: catboost/catboost_pruning.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        self._valid_name = valid_name
+        self._pruned = False
+        self._message = ""
+
+    def after_iteration(self, info: Any) -> bool:
+        """Run after each iteration.
+
+        Args:
+            info:
+                A simpleNamespace containing iteraion, validation_name, metric_name
```
Comment (nzw0301) [inline catboost/catboost_pruning.py:60]: Typo and using inline code:
```suggestion
                A ``simpleNamespace`` containing iteration, ``validation_name``, ``metric_name``
```

File: catboost/catboost_pruning.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                "is not found in the evaluation result list {}.".format(self._valid_name, info)
+            )
+        metrics = info.metrics[self._valid_name]
+        if self._metric not in metrics:
+            raise ValueError(
+                'The entry associated with the metric name "{}" '
+                "is not found in the evaluation result list {}.".format(self._metric, info)
+            )
+        scores = metrics[self._metric]
+        current_score = scores[-1]
```
Comment (nzw0301) [inline catboost/catboost_pruning.py:84]: This is minor thing, but `scores` is only used for `current_score`. How about this simplification?
```suggestion
        current_score = metrics[self._metric][-1]
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Great. I've left minor suggestions. Could you look at them?

File: catboost/catboost_pruning.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    param = {
+        "objective": trial.suggest_categorical("objective", ["Logloss", "CrossEntropy"]),
+        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.01, 0.1, log=True),
+        "depth": trial.suggest_int("depth", 1, 12),
+        "boosting_type": trial.suggest_categorical("boosting_type", ["Ordered", "Plain"]),
+        "bootstrap_type": trial.suggest_categorical(
+            "bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]
+        ),
+        "used_ram_limit": "3gb",
+        "eval_metric": "AUC",
```
Comment (nzw0301) [inline catboost/catboost_pruning.py:111]: Sorry, one more thing. Could you optimize `Accuracy` rather than `AUC`?
```suggestion
        "eval_metric": "Accuracy",
```

File: catboost/catboost_pruning.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        verbose=0,
+        early_stopping_rounds=100,
+        callbacks=[pruning_callback],
+    )
+
+    # evoke pruning manually.
+    pruning_callback.check_pruned()
+
+    preds = gbm.predict(valid_x)
+    auc = roc_auc_score(valid_y, preds)
```
Comment (nzw0301) [inline catboost/catboost_pruning.py:135]: ditto

File: catboost/catboost_pruning.py (hunk 16 above)
Comment (nzw0301) [reply catboost/catboost_pruning.py:111]: I left a comment on this before, so if my comment mislead you, I'm sorry.

**Comment by nzw0301** [comment]
As a follow-up, optimised space can improve by following the castboost's docs: https://catboost.ai/en/docs/concepts/parameter-tuning. In addition, it might be great to unify the code to the native catboost example and the other boosting libraries.

**Comment by tohmae** [comment]
Thank you for your comments.
I reflected your suggestions.

**Review by nzw0301 — APPROVED** [review]
LGTM. Thank you for your contribution 🐱

**Comment by nzw0301** [comment]
Since this PR is a little bit large and different from other examples, I'll wait for other dev members opinions. Please give us a little bit more time.

**Comment by nzw0301** [comment]
I talked with @toshihikoyanase and @HideakiImamura and we were happy to merge this PR into the main branch as a new example! Let me merge this PR. Many thanks, @tohmae!

In addition, we'd be happy to see this `CatBoostPruningCallback` as an integration module of Optuna repo. Note that we need to explain that this `CatBoostPruningCallback` requires `checked_pruned` explicitly unlike other callbacks for pruning without `modify_result` that is introduced by the original PR, https://github.com/optuna/optuna/pull/2734..

I suppose this additional task contradicts my decision in https://github.com/optuna/optuna/pull/2734. I'm really sorry for that.
